flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkBatchProgram.scala

```diff
@@ -292,6 +293,16 @@ object FlinkBatchProgram {
     // convert dynamic partition pruning scan source
     chainedProgram.addLast(DYNAMIC_PARTITION_PRUNING, new FlinkDynamicPartitionPruningProgram)
 
+    // physical rewrite
```

Comment:
@weng1424 thanks for your review. 

It's a good idea. I rechecked the related physical rewrite optimization program in `FlinkBatchProgram`. I think it should also execute in `HEP_RULES_EXECUTION_TYPE.RULE_COLLECTION` not `RULE_SEQUENCE`. In this way, we can remove the extra physical rewrite program. 

After I applied this change, the `PushLocalHashAggIntoScanRule` still did not work as expected because it matched the Exchange -> LocalHashAggregate -> Scan pattern. To make the `PushLocalHashAggIntoScanRule` work as expected, we should also modify the match logic. I will open another ticket to track this

Reply:
https://issues.apache.org/jira/browse/FLINK-31260